### PR TITLE
feat: support issuer aliases

### PR DESCRIPTION
**Pull request by le-yams**
Support OIDC issuer aliases

## Description
Add an option to specify oidc issuer aliases. The aliases are acceptable values for token issuer value (`iss` claim) check.
Note: the token validation still rely on the main issuer jwks.

## References
#1090 

## Review Checklist
- [x] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [x] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

**Comment by rhamzeh** [comment]
Sorry this fell through the cracks @le-yams! We'll have eyes on it soon!

File: cmd/run/run.go [hunk 1]
```diff
@@ -120,6 +120,8 @@ func NewRunCommand() *cobra.Command {
 
 	flags.String("authn-oidc-issuer", defaultConfig.Authn.Issuer, "the OIDC issuer (authorization server) signing the tokens")
 
+	flags.StringSlice("authn-oidc-issuer-aliases", defaultConfig.Authn.IssuerAliases, "the OIDC issuer DNS aliases")
```
Comment (miparnisari) [inline cmd/run/run.go:123]:
```suggestion
	flags.StringSlice("authn-oidc-issuer-aliases", defaultConfig.Authn.IssuerAliases, "the OIDC issuer DNS aliases that will be accepted as valid when verifying tokens")
```

File: cmd/run/run.go [hunk 2]
```diff
@@ -384,7 +386,9 @@ func (s *ServerContext) Run(ctx context.Context, config *serverconfig.Config) er
 		authenticator, err = presharedkey.NewPresharedKeyAuthenticator(config.Authn.Keys)
 	case "oidc":
 		s.Logger.Info("using 'oidc' authentication")
-		authenticator, err = oidc.NewRemoteOidcAuthenticator(config.Authn.Issuer, config.Authn.Audience)
+		issuers := []string{config.Authn.Issuer}
+		issuers = append(issuers, config.Authn.IssuerAliases...)
+		authenticator, err = oidc.NewRemoteOidcAuthenticator(issuers, config.Authn.Audience)
```
Comment (miparnisari) [inline cmd/run/run.go:391]:
```suggestion
		authenticator, err = oidc.NewRemoteOidcAuthenticator(config.Authn.Issuer, config.Authn.IssuerAliases..., config.Authn.Audience)
```

File: internal/authn/oidc/oidc.go [hunk 3]
```diff
@@ -23,8 +24,8 @@ import (
 )
 
 type RemoteOidcAuthenticator struct {
-	IssuerURL string
-	Audience  string
+	IssuerURLs []string
```
Comment (miparnisari) [inline internal/authn/oidc/oidc.go:27]:
```suggestion
        MainIssuer string
	IssuerAliases []string
```

File: internal/authn/oidc/oidc.go [hunk 4]
```diff
@@ -84,7 +85,11 @@ func (oidc *RemoteOidcAuthenticator) Authenticate(requestContext context.Context
 		return nil, errInvalidClaims
 	}
 
-	if ok := claims.VerifyIssuer(oidc.IssuerURL, true); !ok {
+	ok = slices.ContainsFunc(oidc.IssuerURLs, func(issuer string) bool {
```
Comment (miparnisari) [inline internal/authn/oidc/oidc.go:88]:
```suggestion
        validIssuers := []string{oidc.MainIssuer, oidc.IssuerAliases...}
	ok = slices.ContainsFunc(validIssuers, func(issuer string) bool {
```

File: internal/authn/oidc/oidc.go [hunk 5]
```diff
@@ -148,7 +153,7 @@ func (oidc *RemoteOidcAuthenticator) GetKeys() (*keyfunc.JWKS, error) {
 }
 
 func (oidc *RemoteOidcAuthenticator) GetConfiguration() (*authn.OidcConfig, error) {
-	wellKnown := strings.TrimSuffix(oidc.IssuerURL, "/") + "/.well-known/openid-configuration"
+	wellKnown := strings.TrimSuffix(oidc.IssuerURLs[0], "/") + "/.well-known/openid-configuration"
```
Comment (miparnisari) [inline internal/authn/oidc/oidc.go:156]:
```suggestion
	wellKnown := strings.TrimSuffix(oidc.MainIssuer, "/") + "/.well-known/openid-configuration"
```

to make it more obvious

File: cmd/run/run.go [hunk 6]
```diff
@@ -120,6 +120,8 @@ func NewRunCommand() *cobra.Command {
 
 	flags.String("authn-oidc-issuer", defaultConfig.Authn.Issuer, "the OIDC issuer (authorization server) signing the tokens")
```
Comment (miparnisari) [inline cmd/run/run.go:121]:
```suggestion
	flags.String("authn-oidc-issuer", defaultConfig.Authn.Issuer, "the OIDC issuer (authorization server) signing the tokens, and where the keys will be fetched from")
```

File: cmd/run/run_test.go [hunk 7]
```diff
@@ -573,6 +573,80 @@ func TestBuildServerWithOIDCAuthentication(t *testing.T) {
 	}
 }
 
+func TestBuildServerWithOIDCAuthenticationAlias(t *testing.T) {
```
Comment (miparnisari) [inline cmd/run/run_test.go:576]: Once https://github.com/openfga/openfga/pull/1417 is merged, please add unit tests as well.

File: cmd/run/run_test.go [hunk 8]
```diff
@@ -573,6 +573,80 @@ func TestBuildServerWithOIDCAuthentication(t *testing.T) {
...
+		},
+		{
+			_name:      "Missing_header_fails",
+			authHeader: "",
+			expectedErrorResponse: &serverErrors.ErrorResponse{
+				Code:    "bearer_token_missing",
+				Message: "missing bearer token",
+			},
+			expectedStatusCode: 401,
+		},
```
Comment (miparnisari) [inline cmd/run/run_test.go:630]: You don't need these cases `Header_with_invalid_token_fails` and `Missing_header_fails`. These are covered by unit tests and they are not affected by the flags that you are introducing in this PR

File: cmd/run/run_test.go [hunk 9]
```diff
@@ -573,6 +573,80 @@ func TestBuildServerWithOIDCAuthentication(t *testing.T) {
...
+		{
+			_name:              "Correct_token_succeeds",
+			authHeader:         "Bearer " + trustedToken,
+			expectedStatusCode: 200,
+		},
+	}
```
Comment (miparnisari) [inline cmd/run/run_test.go:636]:
```suggestion
		{
			_name:              "Token_with_issuer_equal_to_alias_is_accepted",
			authHeader:         "Bearer " + trustedTokenFromAlias,
			expectedStatusCode: 200,
		},
	}
```

**Review by miparnisari — CHANGES_REQUESTED** [review]

**Comment by le-yams** [comment]
Thank you very much for the review @miparnisari.

> Once https://github.com/openfga/openfga/pull/1417 is merged, please add unit tests as well.

Will do 😃.

File: cmd/run/run_test.go (hunk 7 above)
Comment (miparnisari) [reply cmd/run/run_test.go:570]: It's been merged

Comment (le-yams) [reply cmd/run/run_test.go:570]: @miparnisari if I add a test case to this test suite, would the one I added in `run_test` still be relevant? 
I have the feeling they both would test the same thing. Should I keep the two or only one in `oidc_test` ?

Comment (miparnisari) [reply cmd/run/run_test.go:570]: It's OK to have both, they test the same feature but at different levels. `TestBuildServerWithOIDCAuthenticationAlias` is testing that the server setting `AuthnOIDCConfig` is being passed correctly downstream

Comment (le-yams) [reply cmd/run/run_test.go:570]: Done 😃

**Comment by le-yams** [comment]
The `tests` job failed because of some [rate limiting when uploading the code coverage](https://github.com/openfga/openfga/actions/runs/8251536546/job/22631397221?pr=1354#step:6:55) to Codecov. 

Is it possible to re-run it? 🙏

**Review by miparnisari — APPROVED** [review]

**Pull request merged**